Untagged Album: Add private untagged album

PeterMMM opened the pull request:
### Untagged Album
**Description**
This album functions as a crossover between the "Unsorted" and "Tag" albums, automatically collecting any images that have not been assigned a tag.

Reference #3542

<!--
Thank you for contributing to Lychee! We only accept PR to the master branch.

In addition, please describe the benefit to end users; the reasons it does not break any existing features, etc.

If you're pushing a Feature:
- Title it: "This new feature"
- Describe what the new feature enables
- Add tests to test the new feature.
- Ensure it doesn't break any existing features.

If you're pushing a Fix:
- Title it: "Fixes the bug name"
- If it is a fix an an existing issue start the description with `fixes #xxxx` where xxxx is the issue number.
- Describe how it fixes in a few words.
- Add a test that triggered the bug before fixing it.
- Ensure it doesn't break any feature.

All Pull Requests run with extensive tests for stable and latest versions of PHP. 
Ensure your tests pass or your PR may be taken down.

Additionally you can run `make phpstan` and `make formatting` on your own machine as they will be required to pass for your PR to be accepted.
-->

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added a toggleable "Untagged" Smart Album and album pagination (server-side paginated photo lists).

* **UI**
  * Visible paginators in album view, new "Untagged" badge, and header image selection now supports paginated photo sources.

* **Settings**
  * New enable_untagged toggle and photos-per-page configuration.

* **Documentation**
  * Smart Albums docs and localization entries updated for “Untagged”.

* **Tests**
  * Tests updated to cover the new album type and related validations.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

PeterMMM committed:
Untagged Album: Add private untagged album

This album functions as a crossover between the "Unsorted" and "Tag" albums, automatically collecting any images that have not been assigned a tag.

ildyria commented:
Rather than creating a Model for those, I think it is better to create a SmartAlbum, you can take example on 
https://github.com/LycheeOrg/Lychee/blob/master/app/SmartAlbums/UnsortedAlbum.php
https://github.com/LycheeOrg/Lychee/blob/master/app/SmartAlbums/StarredAlbum.php

Which both extends :
https://github.com/LycheeOrg/Lychee/blob/master/app/SmartAlbums/BaseSmartAlbum.php

That should already simplify some of the work.

Model expects to have a table etc. which would not make sense as there is only one "Untagged album".

ildyria commented on app/Models/UnTaggedAlbum.php:
```diff
@@ -0,0 +1,163 @@
+<?php
```
I assume this one should go as you have the other ones in SmartAlbums. :)

ildyria commented on app/SmartAlbums/UntaggedAlbum.php:
```diff
@@ -0,0 +1,57 @@
+		parent::__construct(
+			id: SmartAlbumType::UNTAGGED,
+			smart_condition: function (Builder $query): void {
+				$query->doesntHave('tags');
```
👍

ildyria commented on app/Relations/HasManyUnTaggedPhotos.php:
```diff
@@ -0,0 +1,71 @@
+<?php
```
Becomes unnecessary since you have the UntaggedAlbum (smart)

ildyria commented:
Thank you for your PR and first draft. 👍 
I like the direction you are taking, I see that you tried to approach and I believe the Smart album one is better.

PeterMMM committed:
Untagged Album: Update untagged album as SmartAlbum, pagination on progress

ildyria commented on app/Http/Resources/Editable/EditableBaseAlbumResource.php:
```diff
@@ -43,7 +43,7 @@ class EditableBaseAlbumResource extends Data
 	public bool $is_model_album;
 	public bool $is_pinned;
 
```
Untagged albums do not have the EditableBaseAlbumResource. That one is used to specify the "description" and other configurations...

And Untagged album is effectively a smart album this one won't be populated.

ildyria commented:
FYI: I am working on a significant refactoring of the front-end to avoid the prop drilling bad practice.
I moved quite a bit of the logic and removed quite a few props (which are now contained in pinia states).
https://github.com/LycheeOrg/Lychee/pull/3721

In think the best/cleanest way to integrate your PR would be by first integrate pagination on all albums,
and then the Unsorted smart albums should be really easy to add as the main pagination logic will be there.
What do you think?

PeterMMM commented:
> FYI: I am working on a significant refactoring of the front-end to avoid the prop drilling bad practice. I moved quite a bit of the logic and removed quite a few props (which are now contained in pinia states). #3721
> 
> In think the best/cleanest way to integrate your PR would be by first integrate pagination on all albums, and then the Unsorted smart albums should be really easy to add as the main pagination logic will be there. What do you think?

@ildyria Well noted for information.
For pagination, I think that's good idea. Keeping pagination logic to common, so we can use pagination to every other albums.

PeterMMM committed:
Untagged Album: Add flexible pagination in untagged album

PeterMMM committed:
Merge remote-tracking branch 'upstream/master' into feature/untagged-album

PeterMMM committed:
Untagged Album: Updated base on Refactor code

PeterMMM committed:
Untagged Album: Revert and remove unneeded changes

PeterMMM committed:
Untagged Album: Refactor code

ildyria commented:
@PeterMMM  Code rabbit reviews are pretty good, don't you think? :)

ildyria commented on resources/js/views/gallery-panels/Album.vue:
```diff
@@ -24,6 +24,11 @@
+		v-model:rows="albumStore.per_page"
+		v-model:first="albumStore.from"
+		@update:rows="(val) => { albumStore.per_page = val; refresh(); }"
+		@update:first="(val) => { albumStore.from = val; refresh(); }"
```
I am not sure about that one, we will need to double check in testing.

PeterMMM commented:
@ildyria Yes, they are quite valid. 🤔 I will check and try to fix them.

PeterMMM committed:
Untagged Album: Update valid recommended changes

ildyria committed:
do not make pagination an exception for untagged, use it globally for all smart albums, prepare road for pagination on ALL albums

ildyria committed:
clean up comments

ildyria commented:
From here it should be also fairly easy to extend the pagination to all albums (but that's for another PR)

ildyria committed:
fix formatting & phpstan

ildyria commented on app/Models/Extensions/SortingDecorator.php:
```diff
@@ -186,4 +187,62 @@ public function get(array $columns = ['*']): Collection
+		try {
+			for ($i = $this->pivot_idx + 1; $i < count($this->order_by); $i++) {
+				$this->base_builder->orderBy($this->order_by[$i]['column'], $this->order_by[$i]['direction']);
+			}
+			// @codeCoverageIgnoreStart
+		} catch (\InvalidArgumentException) {
+			// Sic! In theory, `\InvalidArgumentException` should be thrown
+			// if the *type* of argument differs from the expected type
+			// (e.g. a method gets pass an integer, but requires a string).
+			// If the *value* is invalid, the method should throw a
+			// `\InvalidDomainException`.
+			// But Eloquent throws `\InvalidArgumentException` if the
+			// direction does neither equal "asc" nor "desc".
+			throw new InvalidOrderDirectionException();
+		}
+		// @codeCoverageIgnoreEnd
+
+		/** @var \Illuminate\Pagination\LengthAwarePaginator<int,TModelClass> $result */
+		$result = $this->base_builder->paginate($per_page, $columns, $page_name, $page);
+
+		$items = collect($result->items());
+
+		// Sort with PHP for the remaining criteria in reverse order.
+		for ($i = $this->pivot_idx; $i >= 0; $i--) {
+			$column = $this->order_by[$i]['column'];
+
+			// This conversion is necessary
+			$column_sorting_name = str_replace('photos.', '', $column);
+			$column_sorting_type = ColumnSortingType::tryFrom($column_sorting_name) ?? ColumnSortingType::CREATED_AT;
+
+			$options = in_array($column_sorting_type, self::POSTPONE_COLUMNS, true) ? SORT_NATURAL | SORT_FLAG_CASE : SORT_REGULAR;
+			$items = $items->sortBy(
+				$column_sorting_name,
+				$options,
+				$this->order_by[$i]['direction'] === OrderSortingType::DESC->value
+			)->values();
+		}
+
+		return new LengthAwarePaginator($items, $result->total(), $result->perPage(), $result->currentPage(), $result->getOptions());
+	}
```
Your solution is not a viable solution. Memory use is a problem, hence why we need pagination. It is unfortunate but we will have to bite the bullet and place a disclaimer later in the config.

ildyria committed:
fix tests

ildyria commented on resources/js/stores/AlbumState.ts:
```diff
@@ -32,6 +44,9 @@ export const useAlbumStore = defineStore("album-store", {
 			this.config = undefined;
 			this.isLoading = false;
 		},
+		updateCurrentPage(page: number) {
```
dead code.

ildyria commented on resources/js/stores/AlbumState.ts:
```diff
@@ -18,12 +18,24 @@ export const useAlbumStore = defineStore("album-store", {
 			this.reset();
 			return this.load();
 		},
+		resetPages() {
```
dead code.

ildyria committed:
finishing touches

ildyria committed:
fix tsc

ildyria committed:
Fix tests
